### PR TITLE
Passing random state to pipelines created by IterativeAlgorithm next_batch

### LINKED ISSUE #1181: AutoML: pass random_state to the created pipeline instances
**Problem**
I was digging through the code with @piguy314159265 and [we noticed that `IterativeAlgorithm.next_batch` doesn't initialize pipeline instances with a `random_state` value.](https://github.com/FeatureLabs/evalml/blob/main/evalml/automl/automl_algorithm/iterative_algorithm.py#L60)

**Symptom**
In short, artificially decreased variance in the modeling results.

Suppose we call automl search two times, the first time with the default `random_state` of 0, and the second time with a custom value, like 42. The two calls will generate different data splits, and the automl tuners (after the first batch) could select different pipelines and generate different parameters for those pipelines. That's fine. But, for the first batch at least, the pipelines which use default parameters will run *with the same random seed both times*.

Most users wouldn't notice this. However, it means some of our performance testing results may have higher variance than our results to date have reported. Our perf tests are supposed to run each trial with a different random seed. After this issue is fixed, the pipelines will also run with a different random seed.

**Fix**
[Update IterativeAlgorithm](https://github.com/FeatureLabs/evalml/blob/main/evalml/automl/automl_algorithm/iterative_algorithm.py#L60) to pass down the random seed when creating pipeline instances:
```python
pipeline_class(parameters=self._transform_parameters(pipeline_class, proposed_parameters), random_state=self._random_state)
```

Once we do this, we should do some performance testing to see what the impact is. My recommendation: 20 trials each on the small datasets yaml. Note we also have to resolve [a similar issue in the perf test code](https://github.com/FeatureLabs/evalml_looking_glass/issues/98) before we can test this.

### PR DESCRIPTION
**Pull request by freddyaboulton**
### Pull Request Description
Fixes #1181 . Also noticed a bug where `PipelineBase.clone` uses random_state = 0 as the default - which means the random state was unexpectedly being changed before fit/score.

-----
*After creating the pull request: in order to pass the **release_notes_updated** check you will need to update the "Future Release" section of* `docs/source/release_notes.rst` *to include this pull request by adding :pr:`123`.*

### CODE REVIEW

**Review by jeremyliweishih — COMMENTED** [review]
lgtm!

**Review by jeremyliweishih — APPROVED** [review]
lgtm!

**Review by angela97lin — APPROVED** [review]
Looks good! Based on the original issue, it looks like we still want to do perf testing to see how this changes results--will this be attached to this PR or done separately?

**Comment by freddyaboulton** [comment]
> Looks good! Based on the original issue, it looks like we still want to do perf testing to see how this changes results--will this be attached to this PR or done separately?

Waiting on [looking glass issue 98](https://github.com/FeatureLabs/evalml_looking_glass/issues/98) for perf tests.

**Comment by angela97lin** [comment]
> > Looks good! Based on the original issue, it looks like we still want to do perf testing to see how this changes results--will this be attached to this PR or done separately?
> 
> Waiting on [looking glass issue 98](https://github.com/FeatureLabs/evalml_looking_glass/issues/98) for perf tests but I think we can do them in this PR!

Sweet, thanks @freddyaboulton! 😊

File: evalml/automl/automl_algorithm/iterative_algorithm.py [hunk 1]
```diff
@@ -58,7 +58,7 @@ def next_batch(self):
 
         next_batch = []
         if self._batch_number == 0:
-            next_batch = [pipeline_class(parameters=self._transform_parameters(pipeline_class, {}))
+            next_batch = [pipeline_class(parameters=self._transform_parameters(pipeline_class, {}), random_state=self.random_state)
```
Comment (dsherry) [inline evalml/automl/automl_algorithm/iterative_algorithm.py:61]: Yep!

File: evalml/automl/automl_search.py [hunk 2]
```diff
@@ -644,7 +644,7 @@ def _compute_cv_scores(self, pipeline, X, y):
                 y_threshold_tuning = None
                 if self.optimize_thresholds and self.objective.problem_type == ProblemTypes.BINARY and self.objective.can_optimize_threshold:
                     X_train, X_threshold_tuning, y_train, y_threshold_tuning = train_test_split(X_train, y_train, test_size=0.2, random_state=self.random_state)
-                cv_pipeline = pipeline.clone()
+                cv_pipeline = pipeline.clone(pipeline.random_state)
```
Comment (dsherry) [inline evalml/automl/automl_search.py:647]: Yep

RE your comment in standup this week: @eccabay and I chose this behavior because we felt copying random_state under the hood was dirty. However, in light of our recent conversation about #1324 (component/pipeline determinism), perhaps we should revisit this, and have `clone` copy `random_state` by default. Do you have thoughts @freddyaboulton @eccabay ? We can move the discussion to an issue and resolve there.

File: evalml/tests/automl_tests/test_iterative_algorithm.py [hunk 3]
```diff
@@ -106,6 +107,7 @@ def test_iterative_algorithm_results(ensembling_value, dummy_binary_pipeline_cla
             num_pipelines_classes = (len(dummy_binary_pipeline_classes) + 1) if ensembling_value else len(dummy_binary_pipeline_classes)
             cls = dummy_binary_pipeline_classes[(algo.batch_number - 2) % num_pipelines_classes]
             assert [p.__class__ for p in next_batch] == [cls] * len(next_batch)
+            assert all(check_random_state_equality(p.random_state, algo.random_state) for p in next_batch)
```
Comment (dsherry) [inline evalml/tests/automl_tests/test_iterative_algorithm.py:110]: Very nice

File: evalml/tests/automl_tests/test_automl.py [hunk 4]
```diff
@@ -1450,3 +1452,26 @@ def test_input_not_woodwork_logs_warning(mock_fit, mock_score, caplog, X_y_binar
...
+        def __init__(self, parameters, random_state):
+            random_state = get_random_state(random_state)
+            is_diff_random_state = not check_random_state_equality(random_state, expected_random_state)
+            self.__class__.num_pipelines_different_seed += is_diff_random_state
+            super().__init__(parameters, random_state)
+
+    automl = AutoMLSearch(problem_type="binary", allowed_pipelines=[DummyPipeline],
+                          random_state=expected_random_state)
+    automl.search(X, y)
+    assert DummyPipeline.num_pipelines_different_seed == 0
```
Comment (dsherry) [inline evalml/tests/automl_tests/test_automl.py:1477]: Wow this test is genius 🤯 

Could you add `max_iterations=6` or something in the constructor just to be explicit that we're running more than 1 pipeline? I guess if we were really being paranoid, we'd wanna assert that `DummyPipeline` was initialized more than once too--you could keep a `num_pipelines` alongside `num_pipelines_different_seed`.

**Review by dsherry — APPROVED** [review]
@freddyaboulton this PR rocks!! 🤘 🚀  That unit test is amazing ✨

I left a comment about `clone` and random state but its not important for this PR since you included random state in there. Wanna file an issue and we can discuss there?

File: evalml/automl/automl_search.py (hunk 2 above)
Comment (freddyaboulton) [reply evalml/automl/automl_search.py:647]: Great idea! I filed #1340 .

File: evalml/tests/automl_tests/test_automl.py (hunk 4 above)
Comment (freddyaboulton) [reply evalml/tests/automl_tests/test_automl.py]: Great idea! Nothing wrong with being paranoid 😄

File: evalml/tests/automl_tests/test_automl.py [hunk 5]
```diff
@@ -1450,3 +1452,28 @@ def test_input_not_woodwork_logs_warning(mock_fit, mock_score, caplog, X_y_binar
...
+            random_state = get_random_state(random_state)
+            is_diff_random_state = not check_random_state_equality(random_state, expected_random_state)
+            self.__class__.num_pipelines_init += 1
+            self.__class__.num_pipelines_different_seed += is_diff_random_state
+            super().__init__(parameters, random_state)
+
+    automl = AutoMLSearch(problem_type="binary", allowed_pipelines=[DummyPipeline],
+                          random_state=expected_random_state, max_iterations=10)
+    automl.search(X, y)
+    assert DummyPipeline.num_pipelines_different_seed == 0 and DummyPipeline.num_pipelines_init
```
Comment (dsherry) [inline evalml/tests/automl_tests/test_automl.py:1479]: Great!

**Comment by dsherry** [comment]
@freddyaboulton any reason not to merge this?

**Comment by freddyaboulton** [comment]
@dsherry I was waiting on perf test PR 149 because I was under the impression perf tests were required for this feature!

**Comment by freddyaboulton** [comment]
@dsherry And I discussed and we agree that perf tests are not needed for this feature because there will be no noticeable change in behavior to users (maybe if they are highly overfitting would they notice a change). We'll run the perf tests with different random states for each run in the future but that shouldn't block this fix.
